### PR TITLE
config/types/flannel: added network-config option

### PR DESCRIPTION
**Pull request by dgonyeo**
This commit adds a new field to flannel called network-config. When
contents are provided to it, an `ExecStartPre` is added to the produced
flannel unit that will call `etcdctl` to set flannel's config to the
endpoint that flannel is configured to use.

As an example the following config:

```yaml
flannel:
    version: 0.7.0
    network_config: |-
        { "Network": "10.1.0.0/16" }
    interface: "{PUBLIC_IPV4}"
    etcd_endpoints: "https://127.0.0.1:2379"
    etcd_cafile: "/etc/ssl/etcd/ca.pem"
    etcd_certfile: "/etc/ssl/etcd/cert.pem"
    etcd_keyfile: "/etc/ssl/etcd/key.pem"
```

Generates this unit:
```
[Unit]
Requires=coreos-metadata.service
After=coreos-metadata.service

[Service]
EnvironmentFile=/run/metadata/coreos
Environment="FLANNEL_IMAGE_TAG=v0.7.0"
ExecStart=
ExecStart=/usr/lib/coreos/flannel-wrapper $FLANNEL_OPTS \
  --etcd-endpoints="https://127.0.0.1:2379" \
  --etcd-cafile="/etc/ssl/etcd/ca.pem" \
  --etcd-certfile="/etc/ssl/etcd/cert.pem" \
  --etcd-keyfile="/etc/ssl/etcd/key.pem" \
  --iface="${COREOS_GCE_IP_LOCAL_0}"
ExecStartPre=/usr/bin/etcdctl --endpoints="https://127.0.0.1:2379" --ca-file="/etc/ssl/etcd/ca.pem" --cert-file="/etc/ssl/etcd/cert.pem" --key-file="/etc/ssl/etcd/key.pem" set /coreos.com/network/config "{ \"Network\": \"10.1.0.0/16\" }"

```

Fixes https://github.com/coreos/bugs/issues/1888

### CODE REVIEW

**Comment by dgonyeo** [comment]
Just updated this to handle when tls assets are specified, and to fix the etcdctl command (it was incorrect)

**Comment by dgonyeo** [comment]
review bump

File: doc/configuration.md [hunk 1]
```diff
@@ -1,7 +1,5 @@
 # Configuration Specification #
 
-A Container Linux Configuration, to be processed by ct, is a YAML document conforming to the following specification:
```
Comment (crawford) [inline doc/configuration.md]: Why was this removed?

File: doc/configuration.md [hunk 2]
```diff
@@ -96,6 +94,7 @@ _Note: all fields are optional unless otherwise marked_
   * **_other options_** (string): this section accepts any valid etcd options for the version of etcd specified. For a comprehensive list, please consult etcd's documentation. Note all options here should be in snake_case, not spine-case.
 * **flannel**
   * **version** (string): the version of flannel to be run
+  * **flannel_config** (string): the flannel configuration to write into etcd before flannel starts.
```
Comment (crawford) [inline doc/configuration.md]: to be written

File: doc/configuration.md (hunk 1 above)
Comment (dgonyeo) [inline doc/configuration.md]: I must have mangled the merge conflict here. My bad.

**Comment by dgonyeo** [comment]
Broken into two commits as requested

File: config/types/flannel.go [hunk 3]
```diff
@@ -10,21 +11,38 @@ import (
...
 
 var (
-	ErrFlannelTooOld      = errors.New("invalid flannel version (too old)")
-	ErrFlannelMinorTooNew = errors.New("flannel minor version too new. Only options available in the previous minor version will be supported")
-	OldestFlannelVersion  = *semver.New("0.5.0")
-	FlannelDefaultVersion = *semver.New("0.6.0")
+	ErrFlannelTooOld                  = errors.New("invalid flannel version (too old)")
+	ErrFlannelMinorTooNew             = errors.New("flannel minor version too new. Only options available in the previous minor version will be supported")
+	ErrNetConfigInvalidJSON           = errors.New("flannel network config doesn't appear to be valid JSON")
+	ErrNetConfigProvidedAndKubeMgrSet = errors.New("a flannel network config cannot be provided if kube_subnet_mgr is set")
```
Comment (crawford) [inline config/types/flannel.go]: Drop the "a".

File: config/types/flannel.go [hunk 4]
```diff
@@ -115,6 +143,47 @@ func flannelContents(flannel Flannel, platform string) (string, error) {
 		return "", err
 	}
 
+	if flannel.NetworkConfig != "" {
+		pre := "ExecStartPre=/usr/bin/etcdctl"
+		var endpoints string
+		certs := struct {
+			etcdCAFile   string
```
Comment (crawford) [inline config/types/flannel.go]: Just flatten these to variables.
